doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+>>> versin(x).as_real_imag()
+(-cos(re(x))*cosh(im(x)) + 1, sin(re(x))*sinh(im(x)))
+```
+
+Defining `as_real_imag()` also automatically makes {func}`~.expand_complex`
+work.
+
+```py
+>>> versin(x).expand(complex=True)
+I*sin(re(x))*sinh(im(x)) - cos(re(x))*cosh(im(x)) + 1
+```
+
+#### Miscellaneous `_eval_*` methods
+
+There are many other functions in SymPy whose behavior can be defined on
+custom functions via a custom `_eval_*` method, analogous to the ones
+described above. See the documentation of the specific function for details on
+how define each method.
+
+(custom-functions-complete-examples)=
+## Complete Examples
+
+Here are complete examples for the example functions defined in this guide.
+See the above sections for details on each method.
+
+(custom-functions-versine-full-example)=
+### Versine
+
+The versine (versed sine) function is defined as
+
+$$\operatorname{versin}(x) = 1 - \cos(x).$$
+
+Versine is an example of a simple function defined for all complex numbers.
+The mathematical definition is simple, which makes it straightforward to
+define all the above methods on it (in most cases we can just reuse the
+existing SymPy logic defined on `1 - cos(x)`).
+
+#### Definition
+
+```
+>>> from sympy import Function, cos, expand_trig, Integer, pi, sin
+>>> from sympy.core.logic import fuzzy_and, fuzzy_not
+>>> class versin(Function):
+...     r"""
+...     The versine function.
+...
+...     $\operatorname{versin}(x) = 1 - \cos(x) = 2\sin(x/2)^2.$
+...
+...     Geometrically, given a standard right triangle with angle x in the
+...     unit circle, the versine of x is the positive horizontal distance from
+...     the right angle of the triangle to the rightmost point on the unit
+...     circle. It was historically used as a more numerically accurate way to
+...     compute 1 - cos(x), but it is rarely used today.
+...
+...     References
+...     ==========
+...
+...     .. [1] https://en.wikipedia.org/wiki/Versine
+...     .. [2] https://blogs.scientificamerican.com/roots-of-unity/10-secret-trig-functions-your-math-teachers-never-taught-you/
+...     """
+...     # Define evaluation on basic inputs.
+...     @classmethod
+...     def eval(cls, x):
+...         # If x is an explicit integer multiple of pi, x/pi will cancel and
+...         # be an Integer.
+...         n = x/pi
+...         if isinstance(n, Integer):
+...             return 1 - (-1)**n
+...
+...     # Define numerical evaluation with evalf().
+...     def _eval_evalf(self, prec):
+...         return (2*sin(self.args[0]/2)**2)._eval_evalf(prec)
+...
+...     # Define basic assumptions.
+...     def _eval_is_nonnegative(self):
+...         x = self.args[0]
+...         # versin(x) is nonnegative if x is real
+...         return x.is_real
+...
+...     def _eval_is_positive(self):
+...         x = self.args[0]
+...         # versin(x) is positive if x is real and not an even multiple of pi
+...         return fuzzy_and([x.is_real, fuzzy_not((x/pi).is_even)])
```

Comment:
Creating a new expression like `x/pi` in an assumptions query is problematic. This kind of thing is bad for performance and also often leads to infinite recursion.

Reply:
Hmm, OK. I thought this was a pretty common pattern, but I can see how it might cause problems. 

So is a better pattern to parse the expression (like with the _peeloff_pi helper in the SymPy trig functions)? That seems more complex, especially for end users. What's the simplest way to write this?

Reply:
This kind of thing is more common in the codebase than it should be. There might not really be a better way but there could be some discussion about this in this guide.

The problem is that creating an expression leads to evaluation which leads to assumptions queries. So if assumptions queries create expressions then we have recursion. Whenever you have recursion you need some proof that your recursion terminates. If you know that you are always creating a smaller expression than the one that you started with then that is some proof. Likewise if your assumptions queries only ever query assumptions directly on their args then you know that you are always recursing to a query on a smaller expression tree.

In this case it is probable fine to do this but for a different reason: you are recursing from a query on a `versin` expression to a query on an expression from the set of expressions that do not involve `versin`. If you know that the set of expressions that do not involve `versin` will never create a `versin` then that is safe.

Getting this right is non-trivial and there should be clear warnings about it. The general rule is:

1. Ideally an assumptions query on x is resolved just by doing some direct assumptions queries on `x.args`.
2. Assumption queries should aboid creating any new expressions.
3. If it is necessary to create any new expression then you need to think very carefully about whether this recursion is guaranteed to terminate.

In this particular case you could use `as_independent` to do a structural check that you do actually have a `Mul` involving `pi`:
```
coeff, pi_ = x.as_independent(pi, as_Add=False)
if pi_ is pi:
    return fuzzy_and([x.is_real, fuzzy_not(coeff.is_even)])
elif x.is_real is False:
    return False
```
Here we always know that `as_independent` only creates smaller expressions. At worst it returns the original expression as the coeff and 1 for `pi_`.

Reply:
I see. I do remember seeing issues in the past with that sort of thing. It's not something that we can really show with this example, because `versin` isn't used anywhere else in the codebase, so it's impossible for it to sneak in. But it can happen if someone is editing assumptions for a more core function in SymPy. 

Actually I don't even think it's possible to "think very carefully" about whether the recursion will terminate. It really depends on what assumptions queries are being done elsewhere, and what expressions might create other expressions, which you cannot possibly know unless you've read the whole codebase. It's tough though because there's a lot of advanced logic in other class' assumptions handlers and you really want to be able to reuse that whenever possible.

This really is a major design flaw in the old assumptions. It makes the handlers implement both the specification of what the predicates are and the actual method by which they are resolved. That leads to problems like this where a handler would need to have global information to do that properly, but it can't. It also makes certain optimizations in the inference impossible, even basic ones like not evaluating the same inference twice. Contrast that with the new assumptions where the predicates are specified symbolically and a global solver determines how to resolve everything and which sub-predicates need to be evaluated. Global solvers (like SAT solvers) can handle cyclical inference chains just fine without infinite recursion:

```py
>>> list(satisfiable(And(x >> y, y >> z, z >> x), all_models=True))
[{y: False, z: False, x: False}, {y: True, z: True, x: True}]
```

Reply:
> Actually I don't even think it's possible to "think very carefully" about whether the recursion will terminate

In general no it isn't possible if arbitrary code can be in any of the handlers. However it is possible if we limit what can go into the handlers. Two rules are enough but they need to be applied universally:

1. Assumptions queries should not create new expressions.
2. A query on expr should be resolved only by querying assumptions on the args of expr (and not other queries on expr itself).

With these two rules we always know that any assumptions query only ever recurses to queries on smaller expressions and therefore the recursion must at some point terminate.

> This really is a major design flaw in the old assumptions

The flaw is expecting the old assumptions to be something that they are not. They are relatively well defined for their purpose which is to be a limited inference system for use in evaluation. Remember that much of what they do is not really about "assumptions" in the sense of assuming things about symbols but actually just about computing basic properties often of numeric expressions.